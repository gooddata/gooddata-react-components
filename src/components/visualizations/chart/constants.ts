// (C) 2007-2018 GoodData Corporation
export const VIEW_BY_DIMENSION_INDEX = 1;
export const STACK_BY_DIMENSION_INDEX = 0;
export const PIE_CHART_LIMIT = 20;
export const HEATMAP_DATA_POINTS_LIMIT = 10000;

// for 'view by 2 attributes' or 'grouped attributes'
export const VIEW_BY_ATTRIBUTES_LIMIT = 2;
export const PARENT_ATTRIBUTE_INDEX = 0; // first attribute in 'View By' panel
export const PRIMARY_ATTRIBUTE_INDEX = 1; // second attribute in 'View By' panel
