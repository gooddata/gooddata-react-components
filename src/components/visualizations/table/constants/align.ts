// (C) 2007-2018 GoodData Corporation
export const ALIGN_LEFT = 'left';
export const ALIGN_RIGHT = 'right';
