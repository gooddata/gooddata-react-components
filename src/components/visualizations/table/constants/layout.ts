// (C) 2019 GoodData Corporation
export const DEFAULT_HEADER_HEIGHT: number = 26;
export const DEFAULT_ROW_HEIGHT: number = 30;
export const DEFAULT_FOOTER_ROW_HEIGHT: number = 30;

export const TOTALS_ADD_ROW_HEIGHT: number = 50;
export const TOTALS_TYPES_DROPDOWN_WIDTH: number = 150;
