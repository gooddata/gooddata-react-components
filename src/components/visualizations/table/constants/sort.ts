// (C) 2007-2018 GoodData Corporation
export const ASC = 'asc';
export const DESC = 'desc';
