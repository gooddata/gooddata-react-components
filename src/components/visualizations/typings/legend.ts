// (C) 2007-2020 GoodData Corporation
import { TOP } from "../chart/legend/PositionTypes";

export type PositionType = "left" | "right" | "top" | "bottom" | "auto";

export interface ILegendOptions {
    enabled: boolean;
    toggleEnabled: boolean;
    position: PositionType;
    format: string;
    items: LegendOptionsItemType[];
    responsive?: boolean;
}

export type LegendOptionsItemType = IBaseLegendItem | IHeatmapLegendItem;

export interface IBaseLegendItem {
    name: string;
    color: string; // in format rgb(20,178,226)
    legendIndex: number;
    yAxis: number;
}

export interface IHeatmapLegendItem {
    range: IRange;
    isVisible?: boolean;
    color: string;
    legendIndex: number;
}

export interface IColorLegendItem {
    range: IRange;
    color: string;
}

export interface IGeoChartLegendData {
    colorData?: IColorLegendItem[];
    sizeData?: number[];
}

export interface IRange {
    from: number;
    to: number;
}

export const DEFAULT_LEGEND_CONFIG = {
    enabled: true,
    position: TOP,
};
