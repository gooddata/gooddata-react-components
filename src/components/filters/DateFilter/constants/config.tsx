// (C) 2007-2019 GoodData Corporation
import * as moment from "moment";
import { ExtendedDateFilters } from "@gooddata/typings";

import { platformDateFormat } from "../../../../constants/Platform";

export const defaultDateFilterOptions: ExtendedDateFilters.IDateFilterOptionsByType = {
    allTime: {
        localIdentifier: "ALL_TIME",
        type: "allTime",
        name: "",
        visible: true,
    },
    absoluteForm: {
        localIdentifier: "ABSOLUTE_FORM",
        type: "absoluteForm",
        from: moment()
            .subtract(1, "month")
            .startOf("day")
            .format(platformDateFormat),
        to: moment()
            .startOf("day")
            .format(platformDateFormat),
        name: "",
        visible: true,
    },
    absolutePreset: [],
    relativeForm: {
        localIdentifier: "RELATIVE_FORM",
        type: "relativeForm",
        granularity: "GDC.time.month",
        from: undefined,
        to: undefined,
        name: "",
        visible: true,
        availableGranularities: ["GDC.time.date", "GDC.time.month", "GDC.time.quarter", "GDC.time.year"],
    },
    relativePreset: {
        "GDC.time.date": [
            {
                from: -6,
                to: 0,
                granularity: "GDC.time.date",
                localIdentifier: "LAST_7_DAYS",
                type: "relativePreset",
                visible: true,
                name: "",
            },
            {
                from: -29,
                to: 0,
                granularity: "GDC.time.date",
                localIdentifier: "LAST_30_DAYS",
                type: "relativePreset",
                visible: true,
                name: "",
            },
            {
                from: -89,
                to: 0,
                granularity: "GDC.time.date",
                localIdentifier: "LAST_90_DAYS",
                type: "relativePreset",
                visible: true,
                name: "",
            },
        ],
        "GDC.time.month": [
            {
                from: 0,
                to: 0,
                granularity: "GDC.time.month",
                localIdentifier: "THIS_MONTH",
                type: "relativePreset",
                visible: true,
                name: "",
            },
            {
                from: -1,
                to: -1,
                granularity: "GDC.time.month",
                localIdentifier: "LAST_MONTH",
                type: "relativePreset",
                visible: true,
                name: "",
            },
            {
                from: -11,
                to: 0,
                granularity: "GDC.time.month",
                localIdentifier: "LAST_12_MONTHS",
                type: "relativePreset",
                visible: true,
                name: "",
            },
        ],
        "GDC.time.quarter": [
            {
                from: 0,
                to: 0,
                granularity: "GDC.time.quarter",
                localIdentifier: "THIS_QUARTER",
                type: "relativePreset",
                visible: true,
                name: "",
            },
            {
                from: -1,
                to: -1,
                granularity: "GDC.time.quarter",
                localIdentifier: "LAST_QUARTER",
                type: "relativePreset",
                visible: true,
                name: "",
            },
            {
                from: -3,
                to: 0,
                granularity: "GDC.time.quarter",
                localIdentifier: "LAST_4_QUARTERS",
                type: "relativePreset",
                visible: true,
                name: "",
            },
        ],
        "GDC.time.year": [
            {
                from: 0,
                to: 0,
                granularity: "GDC.time.year",
                localIdentifier: "THIS_YEAR",
                type: "relativePreset",
                visible: true,
                name: "",
            },
            {
                from: -1,
                to: -1,
                granularity: "GDC.time.year",
                localIdentifier: "LAST_YEAR",
                type: "relativePreset",
                visible: true,
                name: "",
            },
        ],
    },
};

export const defaultDateFilterProjectConfig: ExtendedDateFilters.IDateFilterConfigContent = {
    selectedOption: "THIS_MONTH",
    allTime: defaultDateFilterOptions.allTime,
    absoluteForm: defaultDateFilterOptions.absoluteForm,
    relativeForm: {
        // month has to be the first as it should be the default selected option
        granularities: ["GDC.time.month", "GDC.time.date", "GDC.time.quarter", "GDC.time.year"],
        localIdentifier: defaultDateFilterOptions.relativeForm.localIdentifier,
        name: defaultDateFilterOptions.relativeForm.name,
        visible: true,
    },
    relativePresets: Object.keys(defaultDateFilterOptions.relativePreset).reduce(
        (presets: ExtendedDateFilters.IDateFilterRelativePreset[], granularityKey: string) => {
            const granularityPresets: ExtendedDateFilters.IDateFilterRelativePreset[] =
                defaultDateFilterOptions.relativePreset[granularityKey];
            return [...presets, ...granularityPresets];
        },
        [],
    ),
    absolutePresets: [],
};
