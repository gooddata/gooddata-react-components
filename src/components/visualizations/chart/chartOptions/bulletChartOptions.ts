// (C) 2020 GoodData Corporation
import { Execution, VisualizationObject } from "@gooddata/typings";

import { parseValue } from "../../utils/common";
import { IPointData, ISeriesItem } from "../../../../interfaces/Config";
import { unwrap } from "../../../../helpers/utils";
import { IColorStrategy } from "../colorFactory";
import { SECONDARY_MEASURES, TERTIARY_MEASURES } from "../../../../constants/bucketNames";
import { filterOutEmptyBuckets } from "../../../../helpers/mdObjBucketHelper";

export type IBucketLocalIdentifier = VisualizationObject.IBucket["localIdentifier"];

const isComparativeMeasurePresent = (bucketLocalIdentifiers: IBucketLocalIdentifier[]) =>
    bucketLocalIdentifiers.includes(TERTIARY_MEASURES);

const getValue = (
    value: string,
    seriesIndex: number,
    bucketsLocalIdentifiers: IBucketLocalIdentifier[],
): {
    y: number;
    target?: number;
} =>
    isTargetSeries(seriesIndex, bucketsLocalIdentifiers)
        ? {
              target: parseValue(value),
              y: 0,
          }
        : {
              y: parseValue(value),
          };

const getSeriesItemData = (
    seriesItem: string[],
    measureGroup: Execution.IMeasureGroupHeader["measureGroupHeader"],
    seriesIndex: number,
    bucketsLocalIdentifiers: IBucketLocalIdentifier[],
): IPointData[] =>
    seriesItem.map((pointValue: string) => ({
        ...getValue(pointValue, seriesIndex, bucketsLocalIdentifiers),
        format: unwrap(measureGroup.items[seriesIndex]).format,
        marker: {
            enabled: pointValue !== null,
        },
        name: unwrap(measureGroup.items[seriesIndex]).name,
    }));

const getPrimarySeries = (
    seriesItemConfig: IPointData,
    colorStrategy: IColorStrategy,
    comparativeMeasurePresent: boolean,
) => ({
    ...seriesItemConfig,
    color: colorStrategy.getColorByIndex(0),
    pointPadding: comparativeMeasurePresent ? 0.2 : 0.1,
    zIndex: 1,
});

const getTargetSeries = (seriesItemConfig: IPointData, colorStrategy: IColorStrategy) => ({
    ...seriesItemConfig,
    color: colorStrategy.getColorByIndex(1),
    type: "bullet",
    pointPadding: 0,
    targetOptions: {
        width: "100%",
    },
    zIndex: 2,
});

const getComparativeSeries = (seriesItemConfig: IPointData, colorStrategy: IColorStrategy) => ({
    ...seriesItemConfig,
    pointPadding: 0,
    color: colorStrategy.getColorByIndex(2),
    zIndex: 0,
});

const isTargetSeries = (seriesIndex: number, bucketsLocalIdentifiers: IBucketLocalIdentifier[]) =>
    seriesIndex === bucketsLocalIdentifiers.indexOf(SECONDARY_MEASURES);

const isComparativeSeries = (seriesIndex: number, bucketsLocalIdentifiers: IBucketLocalIdentifier[]) =>
    seriesIndex === bucketsLocalIdentifiers.indexOf(TERTIARY_MEASURES);

const getSeries = (
    seriesIndex: number,
    seriesItemConfig: IPointData,
    bucketsLocalIdentifiers: IBucketLocalIdentifier[],
    colorStrategy: IColorStrategy,
) => {
    if (isTargetSeries(seriesIndex, bucketsLocalIdentifiers)) {
        return getTargetSeries(seriesItemConfig, colorStrategy);
    } else if (isComparativeSeries(seriesIndex, bucketsLocalIdentifiers)) {
        return getComparativeSeries(seriesItemConfig, colorStrategy);
    }

    const comparativeMeasurePresent = isComparativeMeasurePresent(bucketsLocalIdentifiers);
    return getPrimarySeries(seriesItemConfig, colorStrategy, comparativeMeasurePresent);
};

export function getBulletChartSeries(
    executionResultData: Execution.DataValue[][],
    measureGroup: Execution.IMeasureGroupHeader["measureGroupHeader"],
    colorStrategy: IColorStrategy,
    buckets: VisualizationObject.IBucket[],
): ISeriesItem[] {
    const notEmptyMeasureBucketsLocalIdentifiers = filterOutEmptyBuckets(buckets).map(
        bucket => bucket.localIdentifier,
    );

    return executionResultData.map((seriesItem: string[], seriesIndex: number) => {
        const seriesItemData = getSeriesItemData(
            seriesItem,
            measureGroup,
            seriesIndex,
            notEmptyMeasureBucketsLocalIdentifiers,
        );

        const seriesItemConfig: IPointData = {
            legendIndex: seriesIndex,
            data: seriesItemData,
            name: measureGroup.items[seriesIndex].measureHeaderItem.name,
        };

        return getSeries(
            seriesIndex,
            seriesItemConfig,
            notEmptyMeasureBucketsLocalIdentifiers,
            colorStrategy,
        );
    });
}
