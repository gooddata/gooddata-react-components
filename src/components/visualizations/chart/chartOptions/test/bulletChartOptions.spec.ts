// (C) 2019-2020 GoodData Corporation
import { Execution, VisualizationObject } from "@gooddata/typings";

import { getBulletChartSeries } from "../bulletChartOptions";
import { BulletChartColorStrategy } from "../../colorFactory";
import { IColorPalette } from "../../../../../interfaces/Config";
import { SECONDARY_MEASURES, MEASURES, TERTIARY_MEASURES } from "../../../../../constants/bucketNames";
import { createMeasure } from "../../../../../../__mocks__/fixtures";

const getMeasureGroupWithItems = (measureGroupItems: any) => ({
    items: measureGroupItems,
});

const getExecutionResponse = (measureGroup: any): Execution.IExecutionResponse => ({
    dimensions: [
        {
            headers: [
                {
                    measureGroupHeader: measureGroup,
                },
            ],
        },
    ],
    links: { executionResult: "foo" },
});

const getColorStrategy = (colorPalette: IColorPalette, measureGroup: any) =>
    new BulletChartColorStrategy(
        colorPalette,
        undefined,
        undefined,
        undefined,
        getExecutionResponse(measureGroup),
        {},
    );

const getExecutionResultForNMeasures = (measuresCount: number) =>
    new Array(measuresCount).fill(undefined).map((_val, index) => [(index + 1) * 100]);

describe("getBulletChartSeries", () => {
    const availableMeasureGroupItems: any = [
        {
            measureHeaderItem: {
                localIdentifier: "m1",
                name: "Primary measure",
                format: "#,##0.00",
            },
        },
        {
            measureHeaderItem: {
                localIdentifier: "m2",
                name: "Target measure",
                format: "#,##0.00",
            },
        },
        {
            measureHeaderItem: {
                localIdentifier: "m3",
                name: "Comparative measure",
                format: "#,##0.00",
            },
        },
    ];

    const colorPaletteRed: IColorPalette = [
        {
            guid: "0",
            fill: {
                r: 255,
                g: 0,
                b: 0,
            },
        },
    ];

    const colorPaletteBlue: IColorPalette = [
        {
            guid: "0",
            fill: {
                r: 0,
                g: 0,
                b: 255,
            },
        },
    ];

    const buckets = [
        {
            localIdentifier: MEASURES,
            items: [createMeasure(1)],
        },
        {
            localIdentifier: SECONDARY_MEASURES,
            items: [createMeasure(2)],
        },
        {
            localIdentifier: TERTIARY_MEASURES,
            items: [createMeasure(3)],
        },
    ];

    it.each([
        [colorPaletteRed, 1, [availableMeasureGroupItems[0]], [buckets[0]]],
        [colorPaletteBlue, 1, [availableMeasureGroupItems[1]], [buckets[1]]],
        [colorPaletteRed, 1, [availableMeasureGroupItems[2]], [buckets[2]]],
        [
            colorPaletteBlue,
            2,
            [availableMeasureGroupItems[0], availableMeasureGroupItems[1]],
            [buckets[0], buckets[1]],
        ],
        [
            colorPaletteRed,
            2,
            [availableMeasureGroupItems[0], availableMeasureGroupItems[2]],
            [buckets[0], buckets[2]],
        ],
        [
            colorPaletteBlue,
            2,
            [availableMeasureGroupItems[1], availableMeasureGroupItems[2]],
            [buckets[1], buckets[2]],
        ],
        [
            colorPaletteRed,
            3,
            [availableMeasureGroupItems[0], availableMeasureGroupItems[1], availableMeasureGroupItems[2]],
            buckets,
        ],
    ])(
        "should return expected bullet chart series",
        (
            colorPalette: IColorPalette,
            numberOfMeasures: number,
            measureGroupItems: any,
            buckets: VisualizationObject.IBucket[],
        ) => {
            const executionResultData = getExecutionResultForNMeasures(numberOfMeasures);
            const measureGroup = getMeasureGroupWithItems(measureGroupItems);
            const colorStrategy = getColorStrategy(colorPalette, measureGroup);
            expect(
                getBulletChartSeries(executionResultData, measureGroup, colorStrategy, buckets),
            ).toMatchSnapshot();
        },
    );
});
