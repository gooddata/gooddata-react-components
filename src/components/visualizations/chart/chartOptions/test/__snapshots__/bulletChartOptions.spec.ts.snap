// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBulletChartSeries should return expected bullet chart series 1`] = `
Array [
  Object {
    "color": "rgb(255,0,0)",
    "data": Array [
      Object {
        "format": "#,##0.00",
        "marker": Object {
          "enabled": true,
        },
        "name": "Primary measure",
        "y": 100,
      },
    ],
    "legendIndex": 0,
    "name": "Primary measure",
    "pointPadding": 0.1,
    "zIndex": 1,
  },
]
`;

exports[`getBulletChartSeries should return expected bullet chart series 2`] = `
Array [
  Object {
    "color": "rgb(0,0,179)",
    "data": Array [
      Object {
        "format": "#,##0.00",
        "marker": Object {
          "enabled": true,
        },
        "name": "Target measure",
        "target": 100,
        "y": 0,
      },
    ],
    "legendIndex": 0,
    "name": "Target measure",
    "pointPadding": 0,
    "targetOptions": Object {
      "width": "100%",
    },
    "type": "bullet",
    "zIndex": 2,
  },
]
`;

exports[`getBulletChartSeries should return expected bullet chart series 3`] = `
Array [
  Object {
    "color": "rgb(148, 161, 173)",
    "data": Array [
      Object {
        "format": "#,##0.00",
        "marker": Object {
          "enabled": true,
        },
        "name": "Comparative measure",
        "y": 100,
      },
    ],
    "legendIndex": 0,
    "name": "Comparative measure",
    "pointPadding": 0,
    "zIndex": 0,
  },
]
`;

exports[`getBulletChartSeries should return expected bullet chart series 4`] = `
Array [
  Object {
    "color": "rgb(0,0,255)",
    "data": Array [
      Object {
        "format": "#,##0.00",
        "marker": Object {
          "enabled": true,
        },
        "name": "Primary measure",
        "y": 100,
      },
    ],
    "legendIndex": 0,
    "name": "Primary measure",
    "pointPadding": 0.1,
    "zIndex": 1,
  },
  Object {
    "color": "rgb(0,0,179)",
    "data": Array [
      Object {
        "format": "#,##0.00",
        "marker": Object {
          "enabled": true,
        },
        "name": "Target measure",
        "target": 200,
        "y": 0,
      },
    ],
    "legendIndex": 1,
    "name": "Target measure",
    "pointPadding": 0,
    "targetOptions": Object {
      "width": "100%",
    },
    "type": "bullet",
    "zIndex": 2,
  },
]
`;

exports[`getBulletChartSeries should return expected bullet chart series 5`] = `
Array [
  Object {
    "color": "rgb(255,0,0)",
    "data": Array [
      Object {
        "format": "#,##0.00",
        "marker": Object {
          "enabled": true,
        },
        "name": "Primary measure",
        "y": 100,
      },
    ],
    "legendIndex": 0,
    "name": "Primary measure",
    "pointPadding": 0.2,
    "zIndex": 1,
  },
  Object {
    "color": "rgb(148, 161, 173)",
    "data": Array [
      Object {
        "format": "#,##0.00",
        "marker": Object {
          "enabled": true,
        },
        "name": "Comparative measure",
        "y": 200,
      },
    ],
    "legendIndex": 1,
    "name": "Comparative measure",
    "pointPadding": 0,
    "zIndex": 0,
  },
]
`;

exports[`getBulletChartSeries should return expected bullet chart series 6`] = `
Array [
  Object {
    "color": "rgb(0,0,179)",
    "data": Array [
      Object {
        "format": "#,##0.00",
        "marker": Object {
          "enabled": true,
        },
        "name": "Target measure",
        "target": 100,
        "y": 0,
      },
    ],
    "legendIndex": 0,
    "name": "Target measure",
    "pointPadding": 0,
    "targetOptions": Object {
      "width": "100%",
    },
    "type": "bullet",
    "zIndex": 2,
  },
  Object {
    "color": "rgb(148, 161, 173)",
    "data": Array [
      Object {
        "format": "#,##0.00",
        "marker": Object {
          "enabled": true,
        },
        "name": "Comparative measure",
        "y": 200,
      },
    ],
    "legendIndex": 1,
    "name": "Comparative measure",
    "pointPadding": 0,
    "zIndex": 0,
  },
]
`;

exports[`getBulletChartSeries should return expected bullet chart series 7`] = `
Array [
  Object {
    "color": "rgb(255,0,0)",
    "data": Array [
      Object {
        "format": "#,##0.00",
        "marker": Object {
          "enabled": true,
        },
        "name": "Primary measure",
        "y": 100,
      },
    ],
    "legendIndex": 0,
    "name": "Primary measure",
    "pointPadding": 0.2,
    "zIndex": 1,
  },
  Object {
    "color": "rgb(179,0,0)",
    "data": Array [
      Object {
        "format": "#,##0.00",
        "marker": Object {
          "enabled": true,
        },
        "name": "Target measure",
        "target": 200,
        "y": 0,
      },
    ],
    "legendIndex": 1,
    "name": "Target measure",
    "pointPadding": 0,
    "targetOptions": Object {
      "width": "100%",
    },
    "type": "bullet",
    "zIndex": 2,
  },
  Object {
    "color": "rgb(148, 161, 173)",
    "data": Array [
      Object {
        "format": "#,##0.00",
        "marker": Object {
          "enabled": true,
        },
        "name": "Comparative measure",
        "y": 300,
      },
    ],
    "legendIndex": 2,
    "name": "Comparative measure",
    "pointPadding": 0,
    "zIndex": 0,
  },
]
`;
